daffodil-runtime1/src/main/scala/org/apache/daffodil/infoset/InfosetImpl.scala

```diff
@@ -971,6 +1006,34 @@ final class DIArray(
   override def isSimple = false
   override def isComplex = false
 
+  // Parsers don't actually call setHidden on DIArrays, only on DIElements. But
+  // DIArrays are always created when there is at least one child element, and
+  // since all children of an array must have the same visibility, we can just
+  // inspect the isHidden property of the first child to determine if the
+  // entire array is hidden.
+  override def isHidden = _contents(0).isHidden
+
+  // Marks/blocks are never taken on arrays, so the infosetWalkerBlockCount
+  // would normally be zero for arrays. The infosetWalkerBlockCount is used to determine
+  // if this array is known to exist or if it could be backtracked. We know
+  // this DIArray exists if the first child is known to exist (i.e. if it has
+  // no blocks).
+  //
+  // Additionally, the infoset walker may delete elements from an array. In
```

Comment:
Agreed, I can try changing this to a Maybe and see if it gets too ugly and see if maybe the object approach might be better.

Reply:
I think this is actually going to require a pretty large refactor to make using Maybe's look clean and maintainable. We have lots of places all over that directly access these children arrays. I suggest we don't make that change as part of this PR.

Any other options? The object approach could work, but then we'd need to override a bunch of functions/val to throw some exception when used. And we sort of already get that via a NullPointerException. Granted NullPointerException feels scarier, but functionally it's the same thing. Thoughts? Any other approaches?

Reply:
Those are the two options I can think of. I expected this would be a big rip and may not want to go into this PR. 
I don't really want to get NPEs, but we can create a ticket for this "improvement"/"cleanup" issue.